File: dart_worker/lib/protocol/protocol.dart
// Request/response protocol types for the JSON-lines stdio worker.
library protocol;

class AnalyzeRequest {
  final int id;
  final String method;
  final Map<String, dynamic> params;

  AnalyzeRequest({required this.id, required this.method, required this.params});

  factory AnalyzeRequest.fromJson(Map<String, dynamic> json) {
    return AnalyzeRequest(
      id: json['id'] as int,
      method: json['method'] as String,
      params: (json['params'] as Map<String, dynamic>?) ?? {},
    );
  }
}

class SymbolInfo {
  final String id;
  final String kind;
  final String name;
  final String file;
  final int line;
  final bool isPrivate;
  final bool isOverride;
  final bool isEntryPoint;
  final bool isWidget;
  final List<String> refs;

  SymbolInfo({
    required this.id,
    required this.kind,
    required this.name,
    required this.file,
    required this.line,
    required this.isPrivate,
    required this.isOverride,
    required this.isEntryPoint,
    required this.isWidget,
    required this.refs,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'kind': kind,
        'name': name,
        'file': file,
        'line': line,
        'is_private': isPrivate,
        'is_override': isOverride,
        'is_entry_point': isEntryPoint,
        'is_widget': isWidget,
        'refs': refs,
      };
}

Map<String, dynamic> successResponse(int id, List<SymbolInfo> symbols) => {
      'id': id,
      'symbols': symbols.map((s) => s.toJson()).toList(),
    };

Map<String, dynamic> errorResponse(int id, String message) => {
      'id': id,
      'error': message,
    };

Map<String, dynamic> pongResponse(int id) => {
      'id': id,
      'pong': true,
    };

File: internal/dart/protocol.go
package dart

// Symbol represents a Dart symbol extracted by the semantic worker.
type Symbol struct {
	// ID is the canonical symbol identifier: "class:lib/foo.dart::MyClass"
	ID string `json:"id"`

	// Kind: class, function, method, variable, typedef, extension
	Kind string `json:"kind"`

	// Name is the symbol name.
	Name string `json:"name"`

	// File is the file path relative to project root.
	File string `json:"file"`

	// Line is the declaration line number.
	Line int `json:"line"`

	// IsPrivate is true for names starting with _.
	IsPrivate bool `json:"is_private"`

	// IsOverride is true for @override annotated methods.
	IsOverride bool `json:"is_override"`

	// IsEntryPoint is true for @pragma('vm:entry-point') symbols.
	IsEntryPoint bool `json:"is_entry_point"`

	// IsWidget is true for StatelessWidget/StatefulWidget subclasses.
	IsWidget bool `json:"is_widget"`

	// IsFrameworkState is true for _FooState classes paired with StatefulWidget Foo.
	IsFrameworkState bool `json:"is_framework_state"`

	// OwnerClass is the enclosing class name for methods (empty for top-level symbols).
	OwnerClass string `json:"owner_class,omitempty"`

	// Refs are file paths where this symbol is referenced.
	Refs []string `json:"refs"`
}

// AnalyzeRequest is the Go side of a JSON-lines request to the Dart worker.
type AnalyzeRequest struct {
	ID     int    `json:"id"`
	Method string `json:"method"`
	Params any    `json:"params"`
}

// AnalyzeParams holds parameters for the "analyze_project" method.
type AnalyzeParams struct {
	Root         string   `json:"root"`
	EntryPoints  []string `json:"entry_points"`
}

// PingParams holds parameters for the "ping" method.
type PingParams struct{}

// AnalyzeResponse is the Go side of a JSON-lines response from the Dart worker.
type AnalyzeResponse struct {
	ID      int       `json:"id"`
	Symbols []*Symbol `json:"symbols,omitempty"`
	Pong    bool      `json:"pong,omitempty"`
	Error   string    `json:"error,omitempty"`
}
